### PR TITLE
[Ready] Cargo Catalog small-order support, new single-pack orders, addition of Ripley

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
Adds the ability for cargo packs to be specified as "small items", indicating that they should be bundled into the same box as other small items (grouped by purchaser). This is mostly because it seems silly to send a whole separate box for each small item.

Adds single-pack versions of insulated gloves, and several armory items. In most cases, the cost-per-item is higher than with the already-existing bundles. The idea here is to offer prices in the range that an individual can hope to reach.

Adds the MK-I Ripley to the catalog, and changes the bounty item to the MK-II.

## Why It's Good For The Game

Gives more of a purpose for economy. Money will be meaningless until there are things that players can use the money for that they actually care about. No one is going to bust their ass to scrape together cash for a scarf, but insulated gloves (or a shotgun) offer a goal worth pursuing.

## Changelog
:cl: Zxaber
tweak: Certain small items purchased through cargo now get grouped into a single box.
add: Added single-order options for several existing products in the Cargo Catalog.
add: Added ALPU MK-I Ripley kit to the cargo catalog, under Engineering.
balance: Cargo Ripley bounty now requires a MK-II.
remove: Removed the Ripley and Odysseus listings in the cargo catalog that were only the control boards and nothing more.
tweak: Ammo and Medkit listings are now single-pack items, and considered small items that get grouped into single boxes. Price per unit for ammo is unchanged, and for medkits is as close to unchanged as is reasonable.
/:cl:

To-do:

- [x]  Make combined-order boxes that include an item requiring access to inherit the access required.
- [x]  Contemplate how bad I had to snowflake the code to make manifests work on combined-order boxes, and possibly move it all to the cargo shuttle code.
- [x] Add more small item listings and fix the merge conflict on packs.dm.
- [x] Possibly make an ~~icon~~ text to indicates orders that will be sent in a combined-order box.

Prices are not set in stone. I want them to be more expensive per-unit, but I also want to be fair about it. Note that these changes shouldn't affect balance for things like arming the crew to fight Nuke Ops, as the cheaper option of purchasing weapons in-bulk has not been changed.

### CODE REVIEW

**Comment by optimumtact** [comment]
pretty good change :+1:

**Comment by zxaber** [comment]
I suppose if any file was gonna conflict, packs.dm is the least annoying one.

**Comment by Angustmeta** [comment]
I'll add any new crates to the wiki. 

The most urgent problem with cargo is Buy Directly imo. QM can order multiple gun crates round start and just open them with one click. This bypasses the whole cracking crates open ordeal. Buy Directly should probably do an ID check dìrectly in the supply (requests) console, and deny unallowed orders.

**Comment by zxaber** [comment]
It's a bit of a toss-up. QM's abusing the department budget isn't anything new, even back before economy.

The problem with requiring department (security) access for personal orders is that assistants aren't going to convince the HoS to open their new laser crate. So you fall right back into it being easier to break into the armory (or just explore space) than to earn the money and buy what you want. You also end up in the situation where people are collecting money but often not getting the things they're buying, which would likewise hamper people bothering with it all.

On the other hand, there already is a list for cargo purchase history. I don't know if it's available to player in-game, but I don't see any reason why the HoP and HoS shouldn't each get the ability to see the purchase history (possibly barring null-crate orders)

**Comment by Tlaltecuhtli** [comment]
they should be in their own category and the prices are still too high for the average person to get em instead of just dragging 3-4 crates to infront of cargo and ask for an epic crate of disablers

**Comment by necromanceranne** [comment]
I agree, might be wise to drop the price further. Some of those can only reasonably be purchased by command, sec and miners getting space cash with their mining points.

**Comment by optimumtact** [comment]
what do you consider average, and how long

**Comment by Tlaltecuhtli** [comment]
you get like 600$ish  in 1h of extended

**Comment by Daxxed** [comment]
As a biased miner who makes about 1000$ in 20 minutes ( not considering loot or buying cash with points ) the prices seem fine.

**Comment by Arianya** [comment]
The fact that a assistant who doesn't take any risks or go out of their way to get money can't afford guns seems more like a feature then a flaw.

Imagine if we had 20 assistants and they all had money for their own personal eguns at 20 minutes. It'd be nations-lite.

**Comment by necromanceranne** [comment]
I mean I wouldn't even consider assistants in this at all but the only easiest method of acquiring this money for just about anyone that I can think of is through mining, space loot and literally robbing people, all of which make this somewhat moot. Especially space loot, which gives you an unreasonable amount of free equipment anyway, some of which is tator gear. And only really one person will benefit from that. There isn't enough methods of generating cash to individuals over department funds right now, which presents a problem in of itself that needs to be fixed.

I agree that a 600 dollar price range is more reasonable for some of the non-weapon crates. Not sure if I would want to reduce weapon prices too much.

**Comment by zxaber** [comment]
I'm certainly not against price adjustments, but I do have some other things I want to add to help economy. In the mean-time, there are ways to coax other crew members into giving you their cash so that you can buy things, but we need worthwhile items for purchase first.

--

- Combo boxes purchased on the cargo budget now inherit the access restrictions of the contents the same way full packs work. This doesn't split items by access, so if you buy a single pair of insulated gloves and also a laser, the gloves also get locked behind security access. If you're buying on the cargo budget, hopefully you've gone through the proper channels. Combo boxes purchased by individuals, just like normal crates, do not have access requirements other than the purchaser's ID; this is the same behavior that already exists when buying weapons packs.

- All manifests are now generated by the shuttle code, rather than the order code. The combo box manifest code needed to be on the shuttle to work, so it was all moved to avoid the nonsense of copied code. No real change in function.

**Comment by zxaber** [comment]
- Small-order boxes are now marked. It's not really noted in the UI exactly what the box means, but it's consistently only on small order items so I hope people can figure it out.
image

I wanted to do it better, but then I learned that the cargo computer uses tgui. I'm not sure how much I really want to wrestle with that. Any ideas on a better indicator?

**Comment by Tlaltecuhtli** [comment]
have their own tab?

**Comment by Angustmeta** [comment]
I like them better in their ordinary categories, so people easier find what they're looking for and realise they have a choice between big and small orders. A new tab might get bloated eventually as new single packs are added.

**Comment by zxaber** [comment]
So I ended up conceding that I'll have to fight with tgui. This is what I've got so far.

image

The purchase buttons are no longer super far to the right (and I don't know if I can push them back over easily), but it now lists which items are "small items", and also items that require access to boot. Thoughts or concerns on this?

**Comment by datorangebottle** [comment]
>adds Ripley MK1 to the catalog

no no no no no no no no no no

You are taking roundstart work away from robotics. Please don't do this.

**Comment by zxaber** [comment]
Robotics is for making borgs.

The point of this is to add things people might actually want to spend cash on. The trouble with most item additions is that they get bound to R&D (because they make sense to). The Ripley being a mech already researched at round start and seemingly having a history behind it existing before the events of the station make it a good candidate for an orderable mech.

For the record, the Ripley kit comes with basic parts. If Cargo asks Robotics to make them a mech, Robo can add a better scanner module and better cap, resulting in a slightly better mech. And anyway, there are other things you can busy yourself with while waiting for posi and borg upgrades research.

**Comment by zxaber** [comment]
So, pack changes:

- Added a small_item pack for the M1911. Currently priced at 2000, which is the same price I have set for a single WT550 rifle. Willing to hear arguments for a different price.
- Added a small_item pack for the M1911 mag. 750, which is likewise the price I set for a single WT550 mag. Full disclosure - there is a space ruin with three 1911s, but only two mags. This will likely affect that, for better or worse.
- Added single_item packs for all four medical kit options. Slightly more expensive per-item than the bundles, but not by much.
- Removed the existing Ripley and Odyssus packs from the Science category. These were only the control boards, and were probably a hold-over from when Robotics used to be able to print all the mech parts without needing tech researched. As is, they're pretty redundant. Honestly, I didn't even know they existed until I stumbled upon the code. The MK-I crate I created still exists in the Engineering category.

As an aside, since I'm adding in single-pack orders of simple items like medkits and magazines, do we need the bundles? For things like gloves and shotguns, the bundles make sense, but for medkits and mags, I don't see much need for both bundles and single-item orders.

**Comment by zxaber** [comment]
Man, I hate merge conflicts.

So the issue I have now is that there's an icon missing from the build because it was added more recently than my fork was made, and fixing the merge conflict caused the code referencing the icon to be in my PR.

What's my easiest fix, here? Do I make a new branch up do date and then copy my changes over?

**Comment by zxaber** [comment]
Right, so I finally got the merge thing removed. I also went ahead and removed the bulk-order options for ammo and medkits. Price-per-unit is unchanged (as close as I could get for the medkits that were 3 for 1000), and it felt redundant to have a listing for a magazine at 750 or four at 3000.

I think I'm ready for review.

**Comment by optimumtact** [comment]
sorry you need to rebuild tgui.js after updating the code (just throw away the conflicted version)

File: code/modules/shuttle/supply.dm [hunk 1]
```diff
@@ -115,13 +159,38 @@ GLOBAL_LIST_INIT(blacklisted_cargo_types, typecacheof(list(
 		SSshuttle.shoppinglist -= SO
 		SSshuttle.orderhistory += SO
 
-		SO.generate(pick_n_take(empty_turfs))
+		if(SO.pack.small_item) //small_item means it gets piled in the miscbox
+			if(SO.paying_account)
+				if(!miscboxes.len || !miscboxes[D.account_holder]) //if there's no miscbox for this person
+					to_chat(world, "DEBUG -- [SO.orderer_ckey] Has no box, as the value [miscboxes[SO.orderer_ckey]] returned false.")
```
Comment (AnturK) [inline code/modules/shuttle/supply.dm]: to_chat(world,

File: code/modules/shuttle/supply.dm [hunk 2]
```diff
@@ -77,7 +77,51 @@ GLOBAL_LIST_INIT(blacklisted_cargo_types, typecacheof(list(
 	if(getDockedId() == "supply_away") // Sell when we get home
 		sell()
 
+/obj/docking_port/mobile/supply/proc/generateManifest(obj/structure/closet/crate/C, var/id, var/ordernum, var/packname) //generates-the-manifests.
```
Comment (AnturK) [inline code/modules/shuttle/supply.dm]: There's already function named like this in supply order, why did you move these, these weren't only used below.

**Review by AnturK — CHANGES_REQUESTED** [review]
There's something inconsistent here with generateManfiest

File: code/modules/shuttle/supply.dm (hunk 2 above)
Comment (zxaber) [reply code/modules/shuttle/supply.dm]: Normally, the order's code does the manifest generation. The problem is that for combo boxes, there's no order code being processed when it comes time to generate the combo manifests. Originally, I had duplicated the manifest code to the shuttle's procs, but I was unhappy with having basically the same function in two different spots for two slightly different boxes, so I moved it all over to the shuttle code.

File: code/modules/shuttle/supply.dm (hunk 1 above)
Comment (zxaber) [reply code/modules/shuttle/supply.dm]: fixed

File: code/modules/shuttle/supply.dm (hunk 2 above)
Comment (AnturK) [reply code/modules/shuttle/supply.dm]: You left the other one intact in here though. Remove it if it's unused then.

File: code/modules/shuttle/supply.dm [hunk 3]
```diff
@@ -115,13 +159,37 @@ GLOBAL_LIST_INIT(blacklisted_cargo_types, typecacheof(list(
...
+					miscboxes["Cargo"].name = "small items crate"
+					miscboxes["Cargo"].req_access = list()
+				for (var/item in SO.pack.contains)
+					new item(miscboxes["Cargo"])
+				if(SO.pack.access)
+					miscboxes["Cargo"].req_access += SO.pack.access
+				misc_order_num["Cargo"] = "[misc_order_num["Cargo"]]#[SO.id]  "
+		else
+			var/obj/structure/closet/crate/crate = SO.generate(pick_n_take(empty_turfs))
+			generateManifest(crate, "", "#[SO.id]", SO.pack.name)
```
Comment (AnturK) [inline code/modules/shuttle/supply.dm]: And you call the same function here, without any grouped or not argument passed in.

File: code/modules/cargo/packs.dm [hunk 4]
```diff
@@ -2315,7 +2332,7 @@
 
 /datum/supply_pack/misc/book_crate
 	name = "Book Crate"
-	desc = "Surplus from the Nanotrasen Archives, these five books are sure to be good reads."
+	desc = "Surplus from the Nanotrasen Archives, These five books are sure to be good reads."
```
[the remark this answers is not in the document]
Comment (zxaber) [reply code/modules/cargo/packs.dm]: I changed that while I was browsing through the catalog looking for anything else that could be changed to single-item orders. I thought it was a period when I first saw it.

I also thought I had reverted that when I later realized it was a comma. I'll fix it when I get the chance to.

[the remark this answers is not in the document]
Comment (zxaber) [reply code/modules/cargo/packs.dm]: Fixed.

File: code/modules/shuttle/supply.dm (hunk 2 above)
Comment (zxaber) [reply code/modules/shuttle/supply.dm]: Moved the code back to supply order, details below. Closing both of these comments

**Comment by zxaber** [comment]
So, code changes:

The shuttle code now makes a list of all the info for each combo box, and then creates new supply order objects to handle the item generation. This means the manifest code can be part of the supply order code again, and so supply pods will work fine.

**Comment by ExcessiveUseOfCobblestone** [comment]
Why do you have a small item column when every small item is called out as such in name?

> Robotics is for making borgs.

And mechs btw.

**Comment by zxaber** [comment]
Mostly for easier readability. Buying a small item with access restrictions on the cargo budget will lock the whole combo box, so it might be nice to know when that purchase of ammo is going to keep you from the medkit.
